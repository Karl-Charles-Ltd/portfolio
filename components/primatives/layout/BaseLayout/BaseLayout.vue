<template>
  <div :class="layoutClasses"><slot /></div>
</template>

<script lang="ts">
/**
 * Base Layout - A grid-like layout system.
 *
 * A BaseLayout must only ever have BaseLayoutItem components as immediate children.
 *
 * The layout object provides us with a column-style layout system. This file
 * contains the basic structural elements, but classes should be complemented
 * with width utilities, for example:
 *
 *   <BaseLayout>
 *     <BaseLayoutItem r="1/2">
 *     </BaseLayoutItem>
 *     <BaseLayoutItem r="1/2">
 *     </BaseLayoutItem>
 *   </BaseLayout>
 *
 * The above will create a two-column structure in which each column will
 * fluidly fill half of the width of the parent. We can have more complex
 * systems:
 *
 *   <BaseLayout>
 *     <BaseLayoutItem desktopSmall="1/3">
 *     </BaseLayoutItem>
 *     <BaseLayoutItem r="1/2" desktopSmall="1/3">
 *     </BaseLayoutItem>
 *     <BaseLayoutItem r="1/2" desktopSmall="1/3">
 *     </BaseLayoutItem>
 *   </BaseLayout>
 *
 * The above will create a system in which the first item will be 100% width
 * until we enter our desktopSmall breakpoint, when it will become 33.333% width. The
 * second and third items will be 50% of their parent, until they also become
 * 33.333% width at the desktopSmall breakpoint.
 *
 * We can also manipulate entire layout systems by adding a series of modifiers
 * to the BaseLayout block. For example:
 *
 *   <BaseLayout reverse>
 *
 * This will reverse the displayed order of the system so that it runs in the
 * opposite order to our source, effectively flipping the system over.
 *
 *   <BaseLayout fill-direction="around">
 *
 * This will cause the system to fill up from the center, putting space around the layout items.
 * Default behaviour is to fill up the layout system from the left, similar to flex-start.
 *
 *   <BaseLayout gutter="medium">
 *
 * This will cause each layout item to have a "medium" gutter size, inline with the spacing units.
 *
 *   <BaseLayout align="stretch">
 *
 * This will cause all the layout items to be of the same height.
 */

import { mixins, Prop, Component } from 'nuxt-property-decorator';
import layoutMixin from '../layoutMixin.vue';

const directions = ['row', 'column'];

@Component
export default class BaseLayout extends mixins(layoutMixin) {
  @Prop({ default: false }) readonly autoWidth!: boolean;
  @Prop({ default: false }) readonly fullWidth!: boolean;
  @Prop({ default: false }) readonly reverseFromSmallDesk!: boolean;
  @Prop({ default: false }) readonly spaceBetweenFromSmallDesk!: boolean;
  @Prop({ default: 'row', validator: (value) => directions.includes(value) }) readonly direction!: string;

  get layoutClasses() {
    return {
      'v-layout': true,
      'v-layout--auto': this.autoWidth,
      'v-layout--full-width': this.fullWidth,
      'v-layout--reverse': this.reverse,
      'v-layout--reverse@smallDesk': this.reverseFromSmallDesk,
      'v-layout--between@smallDesk': this.spaceBetweenFromSmallDesk,
      [`v-layout--${this.align}`]: this.align,
      [`v-layout--${this.direction}`]: this.direction,
      [`v-layout--${this.fillDirection}`]: this.fillDirection,
      [`v-layout--${this.gutter}`]: this.gutter,
    };
  }
}
</script>

<style scoped lang="scss">
@import '~sass-mq';

$flex-layout-gutter-pattern-sizes: (
  '--default': (
    'mobile': $spacing-tiny,
    'tablet': $spacing-base,
    'smallDesk': $spacing-base,
  ),
);

$flex-layout-fixed-gutter-sizes: (
  null: spacing-unit(null),
  '--none': 0,
  '--tiny': $spacing-tiny,
  '--small': $spacing-small,
  '--medium': $spacing-medium,
  '--intermediate': $spacing-intermediate,
  '--large': $spacing-large,
  '--immense': $spacing-immense,
) !default;

$flex-layout-justify-values: (
  '--left': 'flex-start',
  '--center': 'center',
  '--right': 'flex-end',
  '--between': 'space-between',
  '--around': 'space-around',
  '--evenly': 'space-evenly',
) !default;

$flex-layout-alignment-values: (
  '--top': 'flex-start',
  '--middle': 'center',
  '--bottom': 'flex-end',
  '--baseline': 'baseline',
  '--stretch': 'stretch',
) !default;

/* Default/mandatory classes
     ========================================================================== */

.v-layout {
  display: flex;
  flex-flow: row wrap;
  list-style: none;
  padding: 0;
}

/**
 * 1. Cells are full-width and stack on top of each other by default.
 */

::v-deep .v-layout__item {
  flex: 0 1 auto;
  width: 100%; /* [1] */
}

/* Gutters
   ========================================================================== */

/**
 * 1. Add a negative margin on one side only. That way we aren't required to use
 *    a wrapper with the same amount of spacing, though it's common to do so.
 * 2. Neutralize the left margin on `v-layout`.
 */

@each $gutter-size-namespace, $gutter-size in $flex-layout-fixed-gutter-sizes {
  .v-layout#{$gutter-size-namespace} {
    margin-left: -#{$gutter-size}; /* [1] */
    > ::v-deep .v-layout__item {
      padding-left: $gutter-size; /* [2] */
    }
  }
}

/**
 * Responsive pattern gutters
 */

@each $gutter-size-namespace, $gutter-map in $flex-layout-gutter-pattern-sizes {
  .v-layout#{$gutter-size-namespace}-pattern {
    @each $breakpoint, $gutter-size in $gutter-map {
      @include mq($from: $breakpoint) {
        margin-left: -#{$gutter-size}; /* [1] */
        > ::v-deep .v-layout__item {
          padding-left: $gutter-size; /* [2] */
        }
      }
    }
  }
}

/* Automatically size cells by distributing them equally
   ========================================================================== */

.v-layout--auto {
  > ::v-deep .v-layout__item {
    flex: 1 0 0;
  }
}

.v-layout--full-width {
  width: 100%;
}

/* Horizontal alignment
   ========================================================================== */

@each $namespace, $value in $flex-layout-justify-values {
  .v-layout#{$namespace} {
    justify-content: #{$value};
  }
}

.v-layout--between\@smallDesk {
  justify-content: center;
  @include mq($from: smallDesk) {
    justify-content: space-between;
  }
}

/* Vertical alignment
   ========================================================================== */

@each $namespace, $value in $flex-layout-alignment-values {
  .v-layout#{$namespace} {
    align-items: #{$value};
  }

  ::v-deep .v-layout__item#{$namespace} {
    align-self: #{$value};
  }
}

/* Content distribution
   ========================================================================== */

.v-layout--reverse {
  @include mq($from: 'tablet') {
    flex-direction: row-reverse;
  }
}

.v-layout--reverse\@smallDesk {
  @include mq($from: smallDesk) {
    flex-direction: row-reverse;
  }
}

.v-layout--column {
  flex-direction: column;

  &.v-layout--reverse {
    flex-direction: column-reverse;
  }
}
</style>
