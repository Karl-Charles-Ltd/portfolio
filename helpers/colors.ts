/**
 * Get the hex color codes that are used throughout the project.
 * These colors and names should match with
 * assets/scss/settings/_settings.colors.scss.
 *
 * @param {String} colorName - The color name from the theme
 * @returns {String} - The hex code for the matching color
 */
export function colorFromTheme(colorName) {
  const colors = {
    primary: '#370b3c',
    secondary: '#6b2f93',
    tertiary: '#D0438F',
    quaternary: '#f40ee3',
    quinary: '#09ECFD',
    'quinary-light': '#1A5C7E',
    'quinary-dark': '#010226',
    black: '#000000',
    background: '#202227',
    charcoal: '#4D4D4D',
    'grey-one': '#949494',
    'grey-two': '#CDCDCD',
    'light-grey': '#E2E2E2',
    'light-grey-one': '#E2E2E2',
    'light-grey-two': '#F9F9F9',
    white: '#FFFFFF',
  };

  return colors[colorName];
}

export const themeMap = {
  primary: {
    background: 'primary',
    foreground: 'secondary',
  },
  'primary-dark': {
    background: 'primary-dark',
    foreground: 'secondary',
  },
  secondary: {
    background: 'secondary',
    foreground: 'white',
    accent: 'tertiary',
  },
  'secondary-dark': {
    background: 'secondary-dark',
    foreground: 'white',
  },
  'secondary-light': {
    background: 'secondary-light',
    foreground: 'white',
  },
  tertiary: {
    background: 'tertiary',
    foreground: 'secondary',
  },
  quaternary: {
    background: 'quaternary',
    foreground: 'secondary',
  },
  quinary: {
    background: 'quinary',
    foreground: 'white',
  },
  'quinary-darker': {
    background: 'quinary-darker',
    foreground: 'white',
  },
  white: {
    background: 'white',
    foreground: 'secondary',
  },
  charcoal: {
    background: 'charcoal',
    foreground: 'white',
  },
  'light-grey-one': {
    background: 'light-grey-one',
    foreground: 'secondary',
  },
};
