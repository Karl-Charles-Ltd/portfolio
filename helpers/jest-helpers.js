import { shallowMount, createLocalVue } from '@vue/test-utils';
import { createRenderer } from 'vue-server-renderer';
import VueRouter from 'vue-router';

export function setup() {
  const renderer = createRenderer();
  const localVue = createLocalVue();
  const router = new VueRouter();

  // setupPlugins(useVuex, localVue);

  return {
    renderer,
    router,
    localVue,
    shallowMount,
  };
}

export function componentFactory(component, options) {
  return shallowMount(component, options);
}

export function createSnapshot(component, options) {
  const wrapper = componentFactory(component, options);
  expect(wrapper.exists()).toBeTruthy();
  expect(wrapper.html()).toMatchSnapshot();
}
