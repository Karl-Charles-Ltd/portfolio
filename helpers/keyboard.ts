export function addKeyupEventListener(keyCode, callBack) {
  if (!window) return;
  window.addEventListener('keyup', (event) => {
    if (event.keyCode === keyCode) callBack();
  });
}

export function removeKeyupEventListener(keyCode, callBack) {
  if (!window) return;
  window.removeEventListener('keyup', (event) => {
    if (event.keyCode === keyCode) callBack();
  });
}
