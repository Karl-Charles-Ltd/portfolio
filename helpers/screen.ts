import * as screenBreakpoints from '~/constants/screen-breakpoints';

export const breakpoints: string[] = [
  screenBreakpoints.MOBILE,
  screenBreakpoints.TABLET,
  screenBreakpoints.SMALL_DESK,
  screenBreakpoints.LARGE_DESK,
  screenBreakpoints.HUGE_DESK,
];

interface Screen {
  [key: string]: string | null;
}

export default function mq({ from = null, until = null }: Screen, currentScreenSize: string): boolean {
  const currentBreakpointIndex = breakpoints.indexOf(currentScreenSize);

  if (from !== '' && from !== null) {
    const fromBreakpointIndex = breakpoints.indexOf(from || '');
    if (fromBreakpointIndex > currentBreakpointIndex) {
      return false;
    }
  }

  if (until !== '' && until !== null) {
    const untilBreakpointIndex = breakpoints.indexOf(until || '');
    if (untilBreakpointIndex <= currentBreakpointIndex) {
      return false;
    }
  }

  return true;
}
