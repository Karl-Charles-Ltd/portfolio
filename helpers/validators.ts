import { themeMap } from './colors';

export function validateTheme(theme: string, extraColors: string[] = []): boolean {
  const baseThemes: string[] = Object.keys(themeMap);
  const availableThemes: string[] = baseThemes.concat(extraColors);

  return availableThemes.includes(theme);
}
