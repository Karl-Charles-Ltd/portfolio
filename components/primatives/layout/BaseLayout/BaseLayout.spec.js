import BaseLayout from './BaseLayout.vue';
import { setup, componentFactory, createSnapshot } from '~/helpers/jest-helpers';

const { localVue } = setup();

const shallowMountComponent = (options) => componentFactory(BaseLayout, { localVue, ...options });

describe('BaseLayout', () => {
  test('mounts and renders properly', () => {
    createSnapshot(BaseLayout);
  });

  it(`has the auto width class if the autoWidth prop is true`, () => {
    const wrapper = shallowMountComponent({ propsData: { autoWidth: true } });
    expect(wrapper.classes().includes(`v-layout--auto`)).toBeTruthy();
  });

  it(`has the reverse order class if the revers prop is true`, () => {
    const wrapper = shallowMountComponent({ propsData: { reverse: true } });
    expect(wrapper.classes().includes(`v-layout--reverse`)).toBeTruthy();
  });

  const gutters = ['default-pattern', 'none', 'base', 'small', 'medium', 'intermediate', 'large'];
  for (let sizeTest = 0; sizeTest < gutters.length; sizeTest += 1) {
    const gutter = gutters[sizeTest];
    it(`has the ${gutter} gutter class if the size type is ${gutter}`, () => {
      const wrapper = shallowMountComponent({ propsData: { gutter } });
      expect(wrapper.classes().includes(`v-layout--${gutter}`)).toBeTruthy();
    });
  }

  const alignModes = ['top', 'middle', 'bottom', 'baseline', 'stretch'];
  for (let alignModeTest = 0; alignModeTest < alignModes.length; alignModeTest += 1) {
    const alignMode = alignModes[alignModeTest];
    it(`has the ${alignMode} align mode class if the align mode type is ${alignMode}`, () => {
      const wrapper = shallowMountComponent({ propsData: { align: alignMode } });
      expect(wrapper.classes().includes(`v-layout--${alignMode}`)).toBeTruthy();
    });
  }

  const directions = ['row', 'column'];
  for (let directionTest = 0; directionTest < directions.length; directionTest += 1) {
    const direction = directions[directionTest];
    it(`has the ${direction} direction class if the direction type is ${direction}`, () => {
      const wrapper = shallowMountComponent({ propsData: { direction } });
      expect(wrapper.classes().includes(`v-layout--${direction}`)).toBeTruthy();
    });
  }

  const fillDirections = ['left', 'right', 'center', 'around', 'between', 'evenly'];
  for (let fillDirectionsTest = 0; fillDirectionsTest < fillDirections.length; fillDirectionsTest += 1) {
    const fillDirection = fillDirections[fillDirectionsTest];
    it(`has the fill direction ${fillDirection} class if the fill direction type is ${fillDirection}`, () => {
      const wrapper = shallowMountComponent({ propsData: { fillDirection } });
      expect(wrapper.classes().includes(`v-layout--${fillDirection}`)).toBeTruthy();
    });
  }
});
