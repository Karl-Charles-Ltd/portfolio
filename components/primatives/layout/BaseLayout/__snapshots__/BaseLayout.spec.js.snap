// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseLayout mounts and renders properly 1`] = `"<div class=\\"v-layout v-layout--top v-layout--row v-layout--left v-layout--base\\"></div>"`;
